docs/design/specs/runtime-async.md

```diff
@@ -0,0 +1,82 @@
+
+# Runtime-async
+
+Async is currently a feature implemented by various .NET languages as a compiler rewrite to support methods that can "yield" control back to their caller at specific "suspension" points. While effective, it's believed that implementation directly in the .NET runtime can provide improvements, especially in performance.
+
+## Spec modifications
+
+These are proposed modifications to the ECMA-335 specification for runtime-async.
+
+### I.8.4.5 Sync and Async Methods
+
+Methods may be either 'sync' or 'async'. Async methods have a special signature encoding, described in [### I.8.6.1.5 Method signatures].
+
+Sync methods are all other methods.
+
+Unlike sync methods, async methods support suspension. Suspension allows async methods to yield control flow back to their caller at certain well-defined suspension points, and resume execution of the remaining method at a later time or location, potentially on another thread.
+
+Async methods support the following suspension points:
+
+* Calling another async method. No special instructions need to be provided. If the callee suspends, the caller will suspend as well.
+* Using new .NET runtime APIs to "await" an "INotifyCompletion" type. The signatures of these methods shall be:
+  ```C#
+  // public static async2 Task AwaitAwaiterFromRuntimeAsync<TAwaiter>(TAwaiter awaiter) where TAwaiter : INotifyCompletion
+  public static void modreq([System.Runtime]System.Threading.Tasks.Task) AwaitAwaiterFromRuntimeAsync<TAwaiter>(TAwaiter awaiter) where TAwaiter : INotifyCompletion
+  // public static async2 Task UnsafeAwaitAwaiterFromRuntimeAsync<TAwaiter>(TAwaiter awaiter) where TAwaiter : ICriticalNotifyCompletion
+  public static void modreq([System.Runtime]System.Threading.Tasks.Task) UnsafeAwaitAwaiterFromRuntimeAsync<TAwaiter>(TAwaiter awaiter) where TAwaiter : ICriticalNotifyCompletion
+  ```
+
+Each of the above methods will have semantics analogous to the current AsyncTaskMethodBuilder.AwaitOnCompleted/AwaitUnsafeOnCompleted methods. After calling this method, in can be presumed that the task has completed.
+
+Only local variables which are "hoisted" may be used across suspension points. That is, only "hoisted" local variables will have their state preserved after returning from a suspension. On methods with the `localsinit` flag set, non-"hoisted" local variables will be initialized to their default value when resuming from suspension. Otherwise, these variables will have an undefined value. To identify "hoisted" local variables, they must have an optional custom modifier to the `System.Runtime.CompilerServices.HoistedLocal` class, which will be a new .NET runtime API. This custom modifier must be the last custom modifier on the variable. It is invalid for by-ref variables, or variables with a by-ref-like type, to be marked hoisted. Hoisted local variables are stored in managed memory and cannot be converted to unmanaged pointers without explicit pinning.
```

Comment:
```suggestion
Only local variables which are "hoisted" may be used across suspension points. That is, only "hoisted" local variables will have their state preserved after returning from a suspension. Variables not marked as "hoisted" will have an undefined value after suspension points. To identify "hoisted" local variables, they must have an optional custom modifier to the `System.Runtime.CompilerServices.HoistedLocal` class, which will be a new .NET runtime API. This custom modifier must be the last custom modifier on the variable. It is invalid for by-ref variables, or variables with a by-ref-like type, to be marked hoisted. Hoisted local variables are stored in managed memory and cannot be converted to unmanaged pointers without explicit pinning.
```

The prototypes don't do anything like this today and I also don't see how an IL generator can benefit from this guarantee when it only applies after resumption, so I think there's no need to specify it.

Reply:
We should have IL verification rules for this. I think this guarantee is useful for that.

Reply:
What is the verification rule going to be? I do not see how this guarantee helps when the non-resuming case is specified as undefined.

Reply:
The starting point would be the existing verification rules from ECLA-335:

> Verification assumes that the CLI zeroes all memory other than the evaluation stack before it is made visible to programs. A conforming implementation of the CLI shall provide this observable behavior. Furthermore, verifiable methods shall have the localsinit bit set, see Partition II (Flags for Method Headers).

We can work on evolving these rules. 

We have tried to ignore describing the rules for verifiable (ie type and memory safe) IL for a while as we have evolved .NET. We have learned the hard way recently that it might have been a mistake. We do not have clarity about the compiler/runtime contract that we once had, and it leads to confusion.

Reply:
Moving towards a verification rule like "locals not marked hoisted will either have their value from before the suspension point or zero" does not seem very satisfying to me. Requiring the code generators to zero these locals always is possible, but of course comes with extra complexity (and extra bloat after every suspension point in unoptimized codegen). And sadly validating that the values of unhoisted locals are unused across suspension points would be quite an involved verification rule since it requires computing liveness.

As I have expressed before my opinion here is that we should drop the `HoistedLocal` annotation entirely. Then there is no concept of "unhoisted local used across a suspension point", and the use just naturally retains its (potentially zero initialized) value from before the suspension point.

Reply:
OTOH, if we want to do verification that involves byrefs I'm not sure we can avoid the "the verifier needs to compute liveness" problem in any case...

Reply:
I also think explicit `HoistedLocal` is unnecessary concept. JIT will have to infer what is hoisted from usage anyways and will have to reject invalid use anyways - regardless of decoration.  

NOTE: I completely agree with Jan on making the set of rules that is machine-verifiable. I just do not think `HoistedLocal` would help.

I think it will work better if we specify what use of locals are illegal. 

Ideally the rules could be enforced by a reasonably straight-forward analyzer (i.e. ILVerifier). In some cases we may want to be a bit more conservative than we could. 

Being able to do the correctness analysis in one pass would be ideal (if possible). Conversely - if the rules would ask for complex analysis (i.e. iterating to a fixed point), we should ask if what the scenario enables really worth it. 


While JIT could do complex analysis, it should not drive the rules here. The capability of JITs/interpreters with different levels of optimizations or as new features added or with different levels of optimizations.

Reply:
I also think explicit `HoistedLocal` is unnecessary concept. JIT will have to infer what is hoisted from usage anyways and will have to reject invalid use anyways - regardless of decoration.  

NOTE: I completely agree with Jan on making the set of rules that is machine-verifiable. I just do not think `HoistedLocal` would help.

I think it will work better if we specify what use of locals are illegal. 

Ideally the rules could be enforced by a reasonably straight-forward analyzer (i.e. ILVerifier). In some cases we may want to be a bit more conservative than we could. While JIT could do complex analysis, it should not drive the rules here. The capability of JITs/interpreters with different levels of optimizations or as new features added or with different levels of optimizations.

Being able to do the correctness analysis in one pass would be ideal (if possible). 
Conversely - if the rules would ask for complex analysis (i.e. iterating to a fixed point), we should ask if what the scenario enables really worth it. 

Also note that the rules will indirectly inform the rules of high-level languages and we do not want those to be too complex either. C# users should not do liveness analysis in their head. Current async rules do not require that from the user. 
(even though C# compiler does liveness analysis for the purpose of optimal IL generation, it is an internal detail that changed a lot over time - i.e. was completely redone and became more prcesise in Roslyn 1.0)

Reply:
Let's be clear on what motivated my addition of `HoistedLocal` -- making it easier to get the same results as the current C# compiler lowering in the runtime implementation, particularly for interpreters.

I have not heard a lot of discussion on how interpreters should handle the case without HoistedLocal. Let's focus on that.

Reply:
What we need is an overapproximation of the live state at each suspension point. The HoistedLocal mechanism is one such overapproximation, but there are other ones:
1. Just considering all locals live is clearly an overapproximation, and a very simple one. This is what the prototype currently does in tier 0.
1. Live variable analysis provides a much better estimate than the HoistedLocal mechanism since it does not conflate multiple suspension points into one. I do not think it is prohibitively expensive to compute, even for tier 0.
1. If we are scared about super-linearity of live variable analysis then it is possible to approximate liveness by linearizing the CFG, similarly to how linear scan register allocation works.

In my experience the expensive part of (2) is actually looking at the IR, not the super-linear part of computing the liveness sets after. (3) would incur that cost too. (1) and HoistedLocals does not incur this cost of looking at the IR.

Reply:
The other way looking at this - If nearly everything will ignore the decorations, it makes sense for an IL generator to just put `HoistedLocal` on every local that allows it. 
Also for the most accurate description of what is hoistable, besides locals the set might need to include the parameters and `this`. 

I think we can keep `HoistedLocal` as something that we could do, if there is a scenario that can't live without it, but I have doubts so far.

Reply:
So what I'm hearing is that `HoistedLocal` goes too far, but there seems like there's still rules we need to specify.

In particular, does the stack need to be empty before an async call (aside from arguments)?